### PR TITLE
[NN] GCNII model added <https://arxiv.org/pdf/2007.02133.pdf>

### PR DESCRIPTION
**Pull request by kyawlin**
GCNII model added

## Checklist
Please feel free to remove inapplicable items for your PR.
- [X] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [X] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [X] Code is well-documented
- [X] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by VoVAllen** [comment]
To trigger regression tests: 
 - `@dgl-bot run [instance-type] [which tests] [compare-with-branch]`; 
 For example: `@dgl-bot run g4dn.4xlarge all dmlc/master` or `@dgl-bot run c5.9xlarge kernel,api dmlc/master`

**Comment by sneakerkg** [comment]
@VoVAllen May I bother you ask Yongyi to have a check on this PR? I believe GCNII if from their group.

**Review by sneakerkg — CHANGES_REQUESTED** [review]
@kyawlin Thanks for contributing to DGL. I may bother you also adding the module info into: https://github.com/dmlc/dgl/blob/master/docs/source/api/python/nn.pytorch.rst  so that our doc  will automatically build for this module.

File: python/dgl/nn/pytorch/conv/gcn2conv.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        Compute graph convolution.
+
+        Parameters
+        ----------
+        graph : DGLGraph
+            The graph.
+        feat : torch.Tensor
+             it represents the input feature of shape
+            :math:`(N, D_{in})`
+            where :math:`D_{in}` is size of input feature, :math:`N` is the number of nodes.
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/gcn2conv.py:176]:
```suggestion
            The input feature of shape
            :math:`(N, D_{in})`
            where :math:`D_{in}` is the size of input feature and :math:`N` is the number of nodes.
```

File: python/dgl/nn/pytorch/conv/gcn2conv.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        feat_0 : torch.Tensor
+                it represents the initial feature of shape :math:`(N, D_{in})`
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/gcn2conv.py:178]:
```suggestion
            The initial feature of shape :math:`(N, D_{in})`
```

File: python/dgl/nn/pytorch/conv/gcn2conv.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+        Returns
+        -------
+        torch.Tensor
+            The output feature
+
+        Raises
+        ------
+        DGLError
+            Case 1:
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/gcn2conv.py:188]:
```suggestion
```

Or you can add the other cases.

**Review by BarclayII — CHANGES_REQUESTED** [review]

**Comment by kyawlin** [comment]
thanks.  I will look into those. :)

**Comment by BarclayII** [comment]
> @BarclayII could you advice me what I should do next regarding this pull request?

Could you try addressing each of the comments?  Some of the comments have my suggested edits so you can simply click the "Commit suggestion" button if the edits make sense to you.  For other comments you may need manual editing.  Once done you could click the "resolve conversation" button.  You could also reply in the comments directly if you have questions or if the comment does not make sense.

Thanks!

File: python/dgl/nn/pytorch/conv/gcn2conv.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    >>> from dgl.nn import GCN2Conv
+
+    >>> # Homogeneous graph
+    >>> g = dgl.graph(([0,1,2,3,2,5], [1,2,3,4,0,3]))
+    >>> feat = th.ones(6, 3)
+    >>> g = dgl.add_self_loop(g)
+    >>> conv1 = GCN2Conv(3, layer=1, alpha=0.5, \
+                project_initial_features=True, allow_zero_in_degree=True)
+    >>> conv2 = GCN2Conv(3, layer=2, alpha=0.5, \
+                project_initial_features=True, allow_zero_in_degree=True)
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/gcn2conv.py:94]:
```suggestion
    >>> conv1 = GCN2Conv(3, layer=1, alpha=0.5, \
    ...         project_initial_features=True, allow_zero_in_degree=True)
    >>> conv2 = GCN2Conv(3, layer=2, alpha=0.5, \
    ...         project_initial_features=True, allow_zero_in_degree=True)
```

The line continuations should start with `...` like what you see in Python console.

File: python/dgl/nn/pytorch/conv/gcn2conv.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    Description
+    -----------
+    The Graph Convolutional Network via Initial residual
+    and Identity mapping (GCNII) was introduced in `"Simple and Deep Graph Convolutional
+    Networks" <https://arxiv.org/abs/2007.02133>`_ paper.
+    It is mathematically is defined as follows:
+    .. math::
+        \mathbf{h}^{(l+1)} =\left( (1 - \alpha)(\mathbf{D}^{-1/2} \mathbf{\hat{A}}
+        \mathbf{D}^{-1/2})\mathbf{h}^{(l)} + \alpha {\mathbf{h}^{(0)}} \right)
+        \left( (1 - \beta) \mathbf{I} + \beta \mathbf{W} \right)
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/gcn2conv.py:26]:
```suggestion
        \left( (1 - \beta_l) \mathbf{I} + \beta \mathbf{W} \right)
```

File: python/dgl/nn/pytorch/conv/gcn2conv.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+    where :math:`\mathbf{\hat{A}}` is the adjacency matrix with self-loops,
+    :math:`\mathbf{D}_{ii} = \sum_{j=0} \mathbf{A}_{ij}` is its diagonal degree matrix,
+    :math:`\mathbf{h}^{(0)}` is the initial node features,
+    :math:`\mathbf{h}^{(l)}` is the feature of layer :math:`l`,
+    :math:`\alpha` is the fraction of initial node features, and
+    :math:`\beta` is the hyperparameter to tune the strength of identity mapping.
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/gcn2conv.py:33]:
```suggestion
    :math:`\beta_l` is the hyperparameter to tune the strength of identity mapping.
```

**Review by BarclayII — APPROVED** [review]
Awesome!  Just a few more edits and we are good to go.

**Comment by BarclayII** [comment]
Also, two more changes required to really expose your module:

1. Add your module in `docs/source/api/python/nn.pytorch.rst`.  You can just copy-paste one of the entries and change the name to your `GCN2Conv`.
2. Add your module in the `__all__` list in `python/dgl/nn/pytorch/conv/__init__.py`.

That goes for your other PR as well.

**Comment by BarclayII** [comment]
@sneakerkg Do you have more comments?

**Review by sneakerkg — APPROVED** [review]

**Comment by sneakerkg** [comment]
> @sneakerkg Do you have more comments?

LGTM, thanks.

**Pull request merged**